include/revng/EarlyFunctionAnalysis/ApplyRegisterStateDeductions.h

```diff
@@ -0,0 +1,40 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "revng/ADT/SmallMap.h"
+#include "revng/Model/ABI.h"
+#include "revng/Model/Register.h"
+
+namespace abi {
+
+namespace RegisterState {
+
+using namespace model::RegisterState;
+
+} // namespace RegisterState
+
+/// Represents the possibility of the register being used as either
+/// an argument or an return value.
+struct FullRegisterState {
+  RegisterState::Values AsAnArgument = RegisterState::Invalid;
+  RegisterState::Values AsAnReturnValue = RegisterState::Invalid;
+};
+
+/// Stores multiple register states
+using RegisterStateMap = SmallMap<model::Register::Values,
+                                  FullRegisterState,
+                                  16>;
+
+/// Applies abi-based deductions onto a register state map.
+/// For example. If we're sure that the second argument register is used,
+/// then we can also be sure that the first one passes an argument as well
+/// (even if the argument itself is not detected to be used, etc).
+template<model::abi::Values>
+void applyRegisterStateDeductions(RegisterStateMap &State);
```

Comment:
Are you using this in unit tests or something? If not, let's not make this API public.

Reply:
If it's not used, I'd put it in a `.cpp` file where the other overload is defined.

Reply:
Ah, you meant the overload? I thought you were talking about the function in general.

Actually this templated version is the main function - the other one is a slightly slower specialization for the runtime ABI. This first version should suffice for most cases, should it not?

For reference, I'm mostly including the second version since it's an easy way to make sure all the needed template specializations are instantiated without listing them all manually.

Reply:
If you think so I don't mind hiding the "faster" version. Or even straight-up getting rid of it - shouldn't be a complex change.

Reply:
[Done](https://github.com/revng/revng/pull/98/commits/12b57f2bf87db7c6fde78f7c88b733d41116ab92).